### ISSUE TITLE
Installation issue: mysql

### ISSUE THREAD

**Issue by pibion** [spack/spack#15740]
### Spack version
```console
$ spack --version
0.14.1-358-e042bd9
```

### Steps to reproduce the issue

```console
$ spack install <spec> # Fill in the exact spec you are using
... # and the relevant part of the error message
```

### Platform and user environment

Running centos7 with gcc 4.8.5 in a Docker container (https://gist.github.com/pibion/295bef8fb81555d54b81a256f8654f03).

### Additional information

I'm wondering if the issue could be my version of gcc since 

Rebuilding the package with the following options:
```console
$ spack -d install -j 1 mysql
...
```
Prints out the following line in spack-cc-boost-jppggf3.out.log:

```
[cc] /bin/g++ -march=core-avx2 -mtune=core-avx2 -I. -I/home/daq/packages/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/zlib-1.2.11-pkmj6e72vggig3epxjcwxgzmxncaqnmp/include -I/home/daq/packages/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/bzip2-1.0.8-ersrl36bb43nnkjwe4t4gdur5jnwlzye/include -std=c++14 -fvisibility-inlines-hidden -fPIC -m64 -pthread -O3 -finline-functions
-Wno-inline -w -fvisibility=hidden -DBOOST_ALL_DYN_LINK=1 -DBOOST_ALL_NO_LIB=1 -DNDEBUG -c -o bin.v2/libs/wave/build/gcc-4.8.5/release/cxxstd-14-iso/threading-multi/visibility-hidden/cpplexer/re2clex/cpp_re.o libs/wave/src/cpplexer/re2clex/cpp_re.cpp
```
And I remember someone mentioning on this forum that gcc 4.8.5 doesn't support regex.

I'll try updating gcc and report back.

### General information

- [x] I have run `spack --version` and reported the version of Spack
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by pibion** [comment]
Update: I've used Spack to bootstrap to gcc 4.9.4 and the following mysql install command works:

```
spack install mysql@5.7.26%gcc@4.9.4
```

Still having trouble getting my Dockerfile to build but that's probably a separate issue.

It would be nice if boost knew about this compiler dependency and auto-installed it or warned the user.  Is that a boost packaging thing, or a Spack thing?

**Comment by alalazo** [comment]
One possible issue here is that the default C++ standard for `mysql` is `14`, but `gcc@4.8.5`  doesn't support that standard.

**Comment by alalazo** [comment]
Confirmed. The recipe fails building boost with errors like:
```console
$ spack instyall mysql %gcc@4.8.5
[ ... ]
     16268    ...failed gcc.compile.c++ bin.v2/libs/wave/build/gcc-4.8.5/releas
              e/cxxstd-14-iso/threading-multi/visibility-hidden/cpplexer/re2cle
              x/aq.o...
     16269    gcc.compile.c++ bin.v2/libs/wave/build/gcc-4.8.5/release/cxxstd-1
              4-iso/threading-multi/visibility-hidden/cpplexer/re2clex/cpp_re.o
  >> 16270    g++: error: unrecognized command line option '-std=c++14'    
[ ... ]
==> Error: Installation of mysql failed.  Review log for details
```

**Comment by alalazo** [comment]
This can be solved right now with conflicts, or we can wait 0.15.0 for "compiler as dependencies". The advantage with the latter is that language support will be like a virtual dependency, so we need to store the information only once in each package providing a compiler.

**Comment by alalazo** [comment]
Just checked and there are 22 packages as of 8ed8bb0a769e8abc690423eb85c8b53c229a369a that declare the `cxxstd` variant. I think the best way to deal with this is to wait until "compiler as dependencies" becomes a feature. When that happens those packages can declare something like:
```python
depends_on('cxx@14', when=`cxxstd=14`)
```
and the compiler packages will provide the appropriate language support via:
```python
provides('cxx@14', when='@4.9:')
```

**Comment by pibion** [comment]
This sounds reasonable to me, I'm looking forward to the "compiler as dependencies" feature!

Does it make sense to close this issue, or wait until the feature is released and the dependency is added?
